foundry.md

```diff
@@ -145,6 +157,70 @@ This rule then takes the account using `#asWord(#range(LM, ARGSTART +Int 4, 32)`
          </account>
 ```
 
+Pranks
+------
+
+#### `startPrank` - Sets `msg.sender` for all subsequent calls until `stopPrank` is called.
+
+```
+function startPrank(address) external;
+function startPrank(address sender, address origin) external;
+```
+
+`call.startPrank` and `call.startPrankWithOrigin` will match when either of `startPrank` functions are called at the [Foundry cheatcode address](https://book.getfoundry.sh/cheatcodes/#cheatcodes-reference).
+The rule then take the account using `#asWord(#range(LM, ARGSTART +Int 4, 32)` which will be used to impersonate future calls and set them into the `<caller>` and (if is the case) `<origin>` cells.
+The same rules will store the initial values of the `<caller>` and `<origin>` cells in `<prevCaller>` and `<prevOrigin>`.
+
+```k
+    rule [call.startPrank]:
+        <k> CALL _ CHEAT_ADDR 0 ARGSTART _ARGWIDTH 0 0 => 1 ~> #push ... </k>
+        <output> _ => .ByteArray </output>
+        <caller> CL => #asAccount(#range(LM, ARGSTART +Int 4, 32)) </caller>
+        <origin> OG </origin>
+        <prevCaller> _ => CL </prevCaller>
+        <prevOrigin> _ => OG </prevOrigin>
+        <localMem> LM </localMem>
+      requires CHEAT_ADDR ==Int #address(FoundryCheat)
+       andBool #asWord(#range(LM, ARGSTART, 4)) ==Int 105151830 // selector ( "startPrank(address)" )
+      [priority(40)]
+
+    rule [call.startPrankWithOrigin]:
+        <k> CALL _ CHEAT_ADDR 0 ARGSTART _ARGWIDTH 0 0 => 1 ~> #push ... </k>
+        <output> _ => .ByteArray </output>
+        <caller> CL => #asAccount(#range(LM, ARGSTART +Int 4, 32)) </caller>
+        <origin> OG => #asAccount(#range(LM, ARGSTART +Int 36, 32)) </origin>
+        <prevCaller> _ => CL </prevCaller>
+        <prevOrigin> _ => OG </prevOrigin>
+        <localMem> LM </localMem>
+      requires CHEAT_ADDR ==Int #address(FoundryCheat)
+       andBool #asWord(#range(LM, ARGSTART, 4)) ==Int 1169514616 // selector ( "startPrank(address,address)" )
+      [priority(40)]
+```
+
+#### `stopPrank` - stops the virtual machine from impersonating an account.
+
+```
+function stopPrank() external;
+```
+
+`call.stopPrank` will be invoked when the `stopPrank` function is called at the [Foundry cheatcode address](https://book.getfoundry.sh/cheatcodes/#cheatcodes-reference).
+The rule will restore the account addresses stored in `<prevCaller>` and `<prevOrigin>` cells into `<caller>` and `<origin>` cells.
+
+```k
+    rule [call.stopPrank]:
+        <k> CALL _ CHEAT_ADDR 0 ARGSTART _ARGWIDTH 0 0 ~> 1 ~> #push ... </k>
+         <output> _ => .ByteArray </output>
+         <caller> _ => CL </caller>
+         <origin> _ => OG </origin>
+         <prevCaller> CL </prevCaller>
+         <prevOrigin> OG </prevOrigin>
+         <localMem> LM </localMem>
+      requires CHEAT_ADDR ==Int #address(FoundryCheat)
+       andBool CL =/=K .Account andBool OG =/=K .Account
```

Comment:
Why do we need this? Is this a possible scenario?

Reply:
I was thinking about corner case scenarios in which `stopPrank()` is called before a `startPrank`, but I have to double check how foundry actually handles those tests.

Reply:
@anvacaru please write a test which calls stopPrank before startPrank, so we can make sure we handle it similarly to Foundry.

Reply:
Done. Looks like Foundry accepts tests in which `stopPrank` is called before `startPrank`. I've updated the rules accordingly.